File: tools/conformance/src/validators/typescript/files.ts
import fs from "node:fs";
import path from "node:path";

import { workspaceRoot } from "@nx/devkit";

import { converterByStringCase } from "../../constants";
import { StringCase } from "../../types";

import { validateJsonConformance } from "./json/validator";
import { validateMarkdownConformance } from "./markdown/validator";
import { validateTextConformance } from "./text/validator";
import { validateTypescriptConformance } from "./validator";

import type {
  ConformanceError,
  InstanceDirectoryValidationResult,
} from "./types";

const TS_EXTENSIONS = new Set([".cjs", ".js", ".jsx", ".mjs", ".ts", ".tsx"]);

/**
 * Formats the nested results from {@link validateInstancesDirectory} into a
 * single rich multi-section error string, or `null` when all files pass.
 *
 * Errors are grouped by directory and then by file. Each error includes the
 * full paths to both the instance and template files, the error location in
 * each file (line/column or JSON path), the expected value, and a suggested
 * fix — making the output suitable for a coding agent to read and act on.
 */
export function stringifyConformanceErrors(
  results: InstanceDirectoryValidationResult[],
): null | string {
  const directoriesWithErrors = results.filter((result) =>
    result.results.some((fileResult) => fileResult.errors.length > 0),
  );
  if (directoriesWithErrors.length === 0) return null;

  const count = directoriesWithErrors.length;
  const header = `Conformance validation failed — ${String(count)} director${count === 1 ? "y" : "ies"} with errors.`;
  const body = directoriesWithErrors.flatMap(
    ({ directoryName, results: fileResults }, index) =>
      formatDirectoryLines(directoryName, fileResults, index),
  );
  return [header, ...body].join("\n");
}

/**
 * Validates all generated files in a single instance directory against their
 * corresponding Mustache templates in `templateDirectoryPath`.
 *
 * Each template filename may contain `__fieldName__` tokens (e.g.
 * `__nameKebabCase__.service.ts`) that are resolved to the instance filename
 * using the template variable substitutions derived from `instanceDirectoryPath`'s
 * basename. The same substitution map (`nameCamelCase`, `namePascalCase`,
 * `nameSnakeCase`, `nameKebabCase`) is passed to {@link validateInstanceFile}
 * when rendering each template.
 *
 * Validation uses AST superset-checking so that developer modifications to
 * method bodies, constructor arguments, and array contents do not produce
 * false failures — only structurally required template nodes are enforced.
 *
 * file, each carrying the resolved filename and any validation errors.
 */
export function validateInstanceDirectory(args: {
  instanceDirectoryPath: string;
  templateDirectoryPath: string;
}): InstanceDirectoryValidationResult {
  const { instanceDirectoryPath, templateDirectoryPath } = args;
  const name = path.basename(instanceDirectoryPath);
  const data = buildNameData(name);

  const templateFilenames = fs
    .readdirSync(templateDirectoryPath, { withFileTypes: true })
    .filter((node) => node.isFile())
    .map((node) => node.name);

  const results = templateFilenames.map((templateFilename) =>
    resolveTemplateFile({
      data,
      instanceDirectoryPath,
      templateDirectoryPath,
      templateFilename,
    }),
  );

  return { directoryName: name, results };
}

/**
 * File-system variant of {@link validateTypescriptConformance} that reads both the
 * generated instance file and the Mustache template from disk before validating.
 *
 * If either path does not exist (`ENOENT`), returns a structured error with
 * `errorType: 'file'` rather than throwing, so callers can treat a missing file
 * as a conformance failure rather than a crash.
 */
export function validateInstanceFile(args: {
  data: Record<string, unknown>;
  instanceFilePath: string;
  templateFilePath: string;
}): {
  errors: ConformanceError[];
  instanceFilePath: string;
  templateFilePath: string;
} {
  const { data, instanceFilePath, templateFilePath } = args;
  try {
    const instance = fs.readFileSync(instanceFilePath, "utf8");
    const template = fs.readFileSync(templateFilePath, "utf8");
    const filename = path.basename(instanceFilePath);
    const extension = filename.slice(filename.lastIndexOf("."));
    const { errors } = selectValidator({
      data,
      extension,
      filename,
      instance,
      template,
    });
    return { errors, instanceFilePath, templateFilePath };
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code === "ENOENT") {
      return buildMissingFileError(instanceFilePath, templateFilePath);
    }
    throw error;
  }
}

/**
 * Validates every subdirectory of `instancesDirectoryPath` by calling
 * {@link validateInstanceDirectory} on each one.
 *
 * Intended for conformance test suites that need to check an entire
 * `src/modules/` tree (or equivalent) in one call rather than iterating
 * subdirectories manually.
 */
export function validateInstancesDirectory(args: {
  excludeDirectories?: string[];
  instancesDirectoryPath: string;
  templateDirectoryPath: string;
}): InstanceDirectoryValidationResult[] {
  const { excludeDirectories, instancesDirectoryPath, templateDirectoryPath } =
    args;
  return fs
    .readdirSync(instancesDirectoryPath, { withFileTypes: true })
    .filter(
      (node) => node.isDirectory() && !excludeDirectories?.includes(node.name),
    )
    .map((node) =>
      validateInstanceDirectory({
        instanceDirectoryPath: path.join(instancesDirectoryPath, node.name),
        templateDirectoryPath,
      }),
    );
}

/**
 * Build missing file error.
 */
function buildMissingFileError(
  instanceFilePath: string,
  templateFilePath: string,
): {
  errors: ConformanceError[];
  instanceFilePath: string;
  templateFilePath: string;
} {
  return {
    errors: [
      {
        errorType: "file",
        fix: `Create the file using the generator or manually based on the template at ${templateFilePath}`,
        message: `Missing file: ${instanceFilePath}`,
      },
    ],
    instanceFilePath,
    templateFilePath,
  };
}

/**
 * Build name data.
 */
function buildNameData(name: string): Record<string, string> {
  return {
    nameCamelCase: converterByStringCase[StringCase.CAMEL_CASE](name),
    nameKebabCase: converterByStringCase[StringCase.KEBAB_CASE](name),
    namePascalCase: converterByStringCase[StringCase.PASCAL_CASE](name),
    nameSnakeCase: converterByStringCase[StringCase.SNAKE_CASE](name),
  };
}

/**
 * Format directory lines.
 */
function formatDirectoryLines(
  directoryName: string,
  fileResults: InstanceDirectoryValidationResult["results"],
  directoryIndex: number,
): string[] {
  const failingFiles = fileResults.filter((r) => r.errors.length > 0);
  const header = [
    "",
    `${String(directoryIndex + 1)}. directory: ${directoryName}`,
  ];
  const fileLines = failingFiles.flatMap((fileResult, index) =>
    formatFileResultLines(fileResult, index),
  );
  return [...header, ...fileLines];
}

/**
 * Format error lines.
 */
function formatErrorLines(error: ConformanceError, index: number): string[] {
  return [
    "",
    `     ${String(index + 1)}. ${error.message}`,
    ...formatLocationLines({
      column: error.instanceColumn,
      jsonPath: error.instancePath,
      line: error.instanceLine,
      prefix: "Instance",
    }),
    ...formatLocationLines({
      column: error.templateColumn,
      jsonPath: error.templatePath,
      line: error.templateLine,
      prefix: "Template",
    }),
    ...(error.expected === undefined
      ? []
      : [`        Expected: \`${error.expected}\``]),
    ...(error.actual === undefined
      ? []
      : [`        Actual  : \`${error.actual}\``]),
    `        Fix     : ${error.fix}`,
  ];
}

/**
 * Format file result lines.
 */
function formatFileResultLines(
  fileResult: InstanceDirectoryValidationResult["results"][number],
  fileIndex: number,
): string[] {
  const header = [
    "",
    `  ${String(fileIndex + 1)}. file: ${fileResult.filename}`,
    `     Instance: ${path.relative(workspaceRoot, fileResult.instanceFilePath)}`,
    `     Template: ${path.relative(workspaceRoot, fileResult.templateFilePath)}`,
  ];
  const errorLines = fileResult.errors.flatMap((error, index) =>
    formatErrorLines(error, index),
  );
  return [...header, ...errorLines];
}

/**
 * Format location lines.
 */
function formatLocationLines(args: {
  column: number | undefined;
  jsonPath: string | undefined;
  line: number | undefined;
  prefix: string;
}): string[] {
  const { column, jsonPath, line, prefix } = args;
  if (line !== undefined) {
    const column_ = column === undefined ? "" : `, Column ${String(column)}`;
    return [`        ${prefix}: Line ${String(line)}${column_}`];
  }
  if (jsonPath !== undefined) {
    return [`        ${prefix}: JSON path "${jsonPath}"`];
  }
  return [];
}

/**
 * Resolve template file.
 */
function resolveTemplateFile(args: {
  data: Record<string, unknown>;
  instanceDirectoryPath: string;
  templateDirectoryPath: string;
  templateFilename: string;
}): {
  errors: ConformanceError[];
  filename: string;
  instanceFilePath: string;
  templateFilePath: string;
} {
  const {
    data,
    instanceDirectoryPath,
    templateDirectoryPath,
    templateFilename,
  } = args;
  const instanceFilename = templateFilename.replaceAll(
    /__(\w+)__/g,
    (_: string, field: string) => {
      const value = data[field];
      return typeof value === "string" ? value : "";
    },
  );
  const instanceFilePath = path.join(instanceDirectoryPath, instanceFilename);
  const templateFilePath = path.join(templateDirectoryPath, templateFilename);
  return {
    filename: instanceFilename,
    ...validateInstanceFile({ data, instanceFilePath, templateFilePath }),
  };
}

/**
 * Select validator.
 */
function selectValidator(args: {
  data: Record<string, unknown>;
  extension: string;
  filename: string;
  instance: string;
  template: string;
}): { errors: ConformanceError[] } {
  const { data, extension, filename, instance, template } = args;
  if (extension === ".json") {
    return validateJsonConformance({ data, filename, instance, template });
  }
  if (extension === ".md") {
    return validateMarkdownConformance({ data, filename, instance, template });
  }
  if (TS_EXTENSIONS.has(extension)) {
    return validateTypescriptConformance({
      data,
      filename,
      instance,
      template,
    });
  }
  return validateTextConformance({ data, filename, instance, template });
}

File: tools/conformance/src/modules/validator/python/files.py
"""🗂️ File and directory validators."""

import os
from pathlib import Path

from python.constants import converter_by_string_case
from python.notebook.validator import validate_notebook_conformance
from python.types import ConformanceError, StringCase
from python.validator import validate_python_conformance

# Try to use workspace root dynamically, or fallback to current dir
try:
    WORKSPACE_ROOT = Path(__file__).resolve().parent.parent.parent.parent.parent.parent
except Exception:
    WORKSPACE_ROOT = Path.cwd()


def validate_instance_file(
    *, instance_file_path: str | Path, template_file_path: str | Path, data: dict
) -> dict:
    instance_path = Path(instance_file_path)
    template_path = Path(template_file_path)

    if not instance_path.exists():
        rel_path = (
            str(instance_path.relative_to(WORKSPACE_ROOT))
            if instance_path.is_absolute()
            else str(instance_path)
        )
        return {
            "errors": [
                ConformanceError(
                    error_type="file",
                    language="text",
                    message=f'Missing file: "{rel_path}"',
                    fix=f"Create the missing file `{rel_path}` matching the template.",
                )
            ]
        }

    with open(template_path, encoding="utf-8") as f:
        template_text = f.read()

    with open(instance_path, encoding="utf-8") as f:
        instance_text = f.read()

    ext = template_path.suffix.lower()

    if ext == ".ipynb":
        return validate_notebook_conformance(
            data=data,
            filename=instance_path.name,
            instance=instance_text,
            template=template_text,
        )
    elif ext == ".py":
        return validate_python_conformance(
            data=data,
            filename=instance_path.name,
            instance=instance_text,
            template=template_text,
        )

    return {"errors": []}


def _replace_path_variables(path_str: str, data: dict) -> str:
    for key, value in data.items():
        if isinstance(value, str):
            path_str = path_str.replace(f"__{key}__", value)
    return path_str


def validate_instance_directory(
    *, instance_directory_path: str | Path, template_directory_path: str | Path
) -> dict:
    instance_dir = Path(instance_directory_path)
    template_dir = Path(template_directory_path)
    name = instance_dir.name

    description = ""
    pyproject_path = instance_dir / "pyproject.toml"
    if pyproject_path.exists():
        with open(pyproject_path, encoding="utf-8") as f:
            for line in f:
                if line.startswith("description = "):
                    description = line.split("=")[1].strip().strip('"').strip("'")
                    break

    data = {
        "name": name,
        "description": description,
        "nameCamelCase": converter_by_string_case[StringCase.CAMEL_CASE](name),
        "namePascalCase": converter_by_string_case[StringCase.PASCAL_CASE](name),
        "nameSnakeCase": converter_by_string_case[StringCase.SNAKE_CASE](name),
        "nameKebabCase": converter_by_string_case[StringCase.KEBAB_CASE](name),
    }

    results = []

    for root, dirs, files in os.walk(template_dir):
        # Exclude common cache/build directories
        dirs[:] = [
            d
            for d in dirs
            if d not in {".ruff_cache", "__pycache__", ".pytest_cache", ".mypy_cache"}
        ]

        root_path = Path(root)
        rel_root = root_path.relative_to(template_dir)

        rel_instance_root = Path(_replace_path_variables(str(rel_root), data))

        for file in files:
            if file == ".DS_Store":
                continue

            template_file_path = root_path / file
            instance_file_name = _replace_path_variables(file, data)
            instance_file_path = instance_dir / rel_instance_root / instance_file_name

            file_result = validate_instance_file(
                instance_file_path=instance_file_path,
                template_file_path=template_file_path,
                data=data,
            )

            results.append(
                {
                    "filename": instance_file_name,
                    "instanceFilePath": str(instance_file_path),
                    "templateFilePath": str(template_file_path),
                    "errors": file_result.get("errors", []),
                }
            )

    return {"directoryName": name, "results": results}


def validate_instances_directory(
    *,
    instances_directory_path: str | Path,
    template_directory_path: str | Path,
    exclude_directories: list[str] | None = None,
) -> list[dict]:
    instances_dir = Path(instances_directory_path)
    exclude = exclude_directories or []
    results = []

    if not instances_dir.exists():
        return results

    for entry in instances_dir.iterdir():
        if entry.is_dir() and entry.name not in exclude:
            results.append(
                validate_instance_directory(
                    instance_directory_path=entry,
                    template_directory_path=template_directory_path,
                )
            )

    return results


def stringify_conformance_errors(results: list[dict]) -> str | None:
    directories_with_errors = [
        res for res in results if any(f["errors"] for f in res.get("results", []))
    ]

    if not directories_with_errors:
        return None

    lines = []
    dir_count = len(directories_with_errors)
    plural = "y" if dir_count == 1 else "ies"
    lines.append(
        f"Conformance validation failed — {dir_count} director{plural} with errors."
    )

    for dir_index, dir_result in enumerate(directories_with_errors):
        directory_name = dir_result.get("directoryName", "")
        failing_files = [f for f in dir_result.get("results", []) if f["errors"]]

        lines.append("")
        lines.append(f"{dir_index + 1}. directory: {directory_name}")

        for file_index, file_result in enumerate(failing_files):
            instance_path = Path(file_result["instanceFilePath"])
            template_path = Path(file_result["templateFilePath"])

            try:
                rel_instance = instance_path.relative_to(WORKSPACE_ROOT)
            except ValueError:
                rel_instance = instance_path

            try:
                rel_template = template_path.relative_to(WORKSPACE_ROOT)
            except ValueError:
                rel_template = template_path

            lines.append("")
            lines.append(f"  {file_index + 1}. file: {file_result['filename']}")
            lines.append(f"     Instance: {rel_instance}")
            lines.append(f"     Template: {rel_template}")

            for i, err in enumerate(file_result["errors"]):
                lines.append("")
                lines.append(f"     {i + 1}. {err.message}")

                if err.instance_line is not None:
                    col = (
                        f", Column {err.instance_column}"
                        if err.instance_column is not None
                        else ""
                    )
                    lines.append(f"        Instance: Line {err.instance_line}{col}")
                elif err.instance_path is not None:
                    lines.append(f'        Instance: JSON path "{err.instance_path}"')

                if err.template_line is not None:
                    col = (
                        f", Column {err.template_column}"
                        if err.template_column is not None
                        else ""
                    )
                    lines.append(f"        Template: Line {err.template_line}{col}")
                elif err.template_path is not None:
                    lines.append(f'        Template: JSON path "{err.template_path}"')

                if err.expected is not None:
                    lines.append(f"        Expected: `{err.expected}`")
                if err.actual is not None:
                    lines.append(f"        Actual  : `{err.actual}`")
                lines.append(f"        Fix     : {err.fix}")

    return "\n".join(lines)
